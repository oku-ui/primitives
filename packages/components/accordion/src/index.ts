export { OkuAccordion } from './accordion'

export {
  type AccordionMultipleEmits, accordionProps, type AccordionMultipleProps, type AccordionSingleEmits, type AccordionSingleProps,
  accordionContentProps,
  type AccordionContentProps, accordionHeaderProps,
  type AccordionHeaderProps, accordionImplProps,
  type AccordionImplProps, type AccordionImplEmits,
  accordionImplSingleProps,
  type AccordionImplSingleProps, type AccordionImplSingleEmits,
  accordionImplMultipleProps,
  type AccordionImplMultipleProps, type AccordionImplMultipleEmits,
  accordionItemProps,
  type AccordionItemProps, type AccordionItemEmits,
  accordionTriggerProps,
  type AccordionTriggerProps, type AccordionTriggerEmits,
} from './props'
export {
  OkuAccordionContent,
} from './accordionContent'

export {
  OkuAccordionHeader,
} from './accordionHeader'

export {
  OkuAccordionImpl,
} from './accordionImpl'

export {
  OkuAccordionImplSingle,
} from './accordionImplSingle'

export {
  OkuAccordionImplMultiple,
} from './accordionImplMultiple'

export {
  OkuAccordionItem,
} from './accordionItem'

export {
  OkuAccordionTrigger,
} from './accordionTrigger'
