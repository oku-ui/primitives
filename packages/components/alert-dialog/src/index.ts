export {
  OkuAlertDialog,
} from './alert-dialog'

export {
  OkuAlertDialogAction,
} from './alert-dialog-action'

export {
  OkuAlertDialogCancel,
} from './alert-dialog-cancel'

export {
  OkuAlertDialogContent,
} from './alert-dialog-content'

export {
  OkuAlertDialogDescription, alertDialogDescriptionProps,
} from './alert-dialog-description'

export {
  OkuAlertDialogDescriptionWarning, dialogDescriptionWarningProps,
} from './alert-dialog-description-warning'

export {
  OkuAlertDialogOverlay, alertDialogOverlayProps,
} from './alert-dialog-overlay'

export {
  OkuAlertDialogPortal, alertDialogPortalProps,
} from './alert-dialog-portal'

export {
  OkuAlertDialogTitle, alertDialogTitleProps,
} from './alert-dialog-title'

export {
  OkuAlertDialogTrigger, alertDialogTriggerProps,
} from './alert-dialog-trigger'

export type {
  AlertDialogProps, AlertDialogActionProps, AlertDialogCancelProps, AlertDialogContentProps, AlertDialogDescriptionProps, AlertDialogOverlayProps, AlertDialogPortalProps, AlertDialogTitleProps, AlertDialogTriggerProps, DescriptionWarningProps,
} from './props'

export {
  alertDialogProps,
  alertDialogActionProps,
  alertDialogCancelProps,
  alertDialogContentProps,
} from './props'
