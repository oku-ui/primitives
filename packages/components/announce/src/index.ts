export {
  OkuAnnounce,
} from './announce'

export type { AnnounceProps, AnnounceElement } from './announce'
