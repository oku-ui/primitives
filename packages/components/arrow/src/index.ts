export {
  OkuArrow,
  arrowProps,
} from './arrow'

export type {
  ArrowProps,
  ArrowElement,
  ArrowIntrinsicElement,
} from './arrow'
