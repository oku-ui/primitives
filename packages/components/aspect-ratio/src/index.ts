export { OkuAspectRatio } from './aspect-ratio'
export type {
  AspectRatioProps,
  AspectRatioElement,
  AspectRatioIntrinsicElement,
} from './aspect-ratio'
