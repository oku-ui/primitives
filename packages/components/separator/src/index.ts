export {
  OkuSeparator,
  separatorProps,
} from './separator'

export type {
  SeparatorProps,
  SeparatorElement,
  SeparatorNaviteElement,
} from './separator'
