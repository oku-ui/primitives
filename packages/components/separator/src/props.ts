import type { PropType } from 'vue'
import { primitiveProps } from '@oku-ui/primitive'
import type { OkuElement, PrimitiveProps } from '@oku-ui/primitive'
import { getInvalidOrientationError, isValidOrientation } from './utils'

export const SEPARATOR_NAME = 'OkuSeparator'

/* -------------------------------------------------------------------------------------------------
 *  Separator - separator.ts
 * ----------------------------------------------------------------------------------------------- */

export type SeparatorNativeElement = OkuElement<'div'>
export type SeparatorElement = HTMLDivElement

export const DEFAULT_ORIENTATION = 'horizontal'
export const ORIENTATIONS = ['horizontal', 'vertical'] as const

export type Orientation = typeof ORIENTATIONS[number]

export interface SeparatorProps extends PrimitiveProps {
  /**
   * Either `vertical` or `horizontal`. Defaults to `horizontal`.
   */
  orientation?: Orientation
  /**
   * Whether or not the component is purely decorative. When true, accessibility-related attributes
   * are updated so that that the rendered element is removed from the accessibility tree.
   */
  decorative?: boolean
}

export const separatorProps = {
  props: {
    orientation: {
      type: String as PropType<SeparatorProps['orientation']>,
      default: DEFAULT_ORIENTATION,
      validator: (value: string): boolean => {
        if (!isValidOrientation(value))
          console.error(getInvalidOrientationError(String(value), SEPARATOR_NAME))

        return true
      },
    },
    decorative: {
      type: Boolean as PropType<SeparatorProps['decorative']>,
    },
    ...primitiveProps,
  },
  emits: { },
}
