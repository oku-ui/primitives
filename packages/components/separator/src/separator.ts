import type { PropType } from 'vue'
import { computed, defineComponent, h, mergeProps, reactive, toRefs } from 'vue'
import type { OkuElement, PrimitiveProps } from '@oku-ui/primitive'
import { Primitive, primitiveProps } from '@oku-ui/primitive'
import { reactiveOmit, useForwardRef } from '@oku-ui/use-composable'

const NAME = 'OkuSeparator'
const DEFAULT_ORIENTATION = 'horizontal'
const ORIENTATIONS = ['horizontal', 'vertical'] as const

type Orientation = typeof ORIENTATIONS[number]

export type SeparatorNaviteElement = OkuElement<'div'>
export type SeparatorElement = HTMLDivElement

export interface SeparatorProps extends PrimitiveProps {
  /**
   * Whether or not the component is purely decorative. When true, accessibility-related attributes
   * are updated so that that the rendered element is removed from the accessibility tree.
   */
  decorative?: boolean
  /**
   * Either `vertical` or `horizontal`. Defaults to `horizontal`.
   */
  orientation?: Orientation
}

export const separatorProps = {
  props: {
    /**
     * Whether or not the component is purely decorative. When true, accessibility-related attributes
     * are updated so that that the rendered element is removed from the accessibility tree.
     */
    decorative: {
      type: Boolean as PropType<boolean | undefined>,
      default: undefined,
    },
    /**
     * Either `vertical` or `horizontal`. Defaults to `horizontal`.
     */
    orientation: {
      type: String as PropType<Orientation | undefined>,
      default: undefined,
    },
    ...primitiveProps,
  },
}

const separator = defineComponent({
  name: NAME,
  inheritAttrs: false,
  props: {
    ...separatorProps.props,
  },
  setup(props, { attrs, slots }) {
    const { decorative, orientation: asOrientation, ...domProps } = toRefs(props)
    const _reactive = reactive(domProps)
    const reactiveDomProps = reactiveOmit(_reactive, (key, _value) => key === undefined)

    const orientation = computed(() => ORIENTATIONS.includes(asOrientation.value!) ? asOrientation.value : DEFAULT_ORIENTATION)
    // `aria-orientation` defaults to `horizontal` so we only need it if `orientation` is vertical
    const ariaOrientation = computed(() => orientation.value === 'vertical' ? orientation.value : undefined)
    const semanticProps = computed(() => decorative.value ? { role: 'none' } : { 'aria-orientation': ariaOrientation?.value, 'role': 'separator' })

    const forwardedRef = useForwardRef()

    const originalReturn = () => {
      if (!isValidOrientation(orientation.value))
        return new Error(getInvalidOrientationError(String(orientation.value), NAME))

      return h(
        Primitive.div,
        {
          'data-orientation': orientation.value,
          ...semanticProps.value,
          ...mergeProps(attrs, reactiveDomProps),
          'ref': forwardedRef,
        },
        {
          default: () => slots.default?.(),
        },
      )
    }

    return originalReturn
  },
})

// Split this out for clearer readability of the error message.
function getInvalidOrientationError(value: string, componentName: string) {
  return `Invalid prop \`orientation\` of value \`${value}\` supplied to \`${componentName}\`, expected one of:
  - horizontal
  - vertical

Defaulting to \`${DEFAULT_ORIENTATION}\`.`
}

function isValidOrientation(orientation: any): orientation is Orientation {
  return ORIENTATIONS.includes(orientation)
}

// TODO: https://github.com/vuejs/core/pull/7444 after delete
export const OkuSeparator = separator as typeof separator &
(new () => {
  $props: SeparatorNaviteElement
})
