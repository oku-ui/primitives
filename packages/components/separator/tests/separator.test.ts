import { describe, expect, it } from 'vitest'
import { mount } from '@vue/test-utils'
import type { Component } from 'vue'
import { h } from 'vue'
import { OkuSeparator } from '../src'

const component = {
  setup(props, { attrs, slots }) {
    return () => h(OkuSeparator, { ...attrs }, slots)
  },
} as Component

describe('okuSeparator', () => {
  const wrapper = mount(component)

  it('renders correctly', async () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it('sets role as separator without decorative', async () => {
    expect(wrapper.attributes('role')).toBe('separator')
  })

  it('sets default orientation horizontal', async () => {
    expect(wrapper.attributes('data-orientation')).toBe('horizontal')
  })

  it('sets role as none on adding decorative', async () => {
    const wrapper = mount(component, {
      propsData: {
        decorative: true,
      },
    })
    expect(wrapper.attributes('role')).toBe('none')
  })

  it('sets orientation vertical', async () => {
    const wrapper = mount(component, {
      propsData: {
        orientation: 'vertical',
      },
    })
    expect(wrapper.attributes('data-orientation')).toBe('vertical')
  })
})
