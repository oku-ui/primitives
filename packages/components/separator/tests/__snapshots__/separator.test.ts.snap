// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OkuSeparator > renders correctly 1`] = `
<div
  data-orientation="horizontal"
  role="separator"
>
  <!---->
</div>
`;
